### ISSUE TITLE
Make contact name configurable

### ISSUE THREAD

**Issue by inoric** [onsip/SIP.js#791]
**Is your feature request related to a problem? Please describe.**
The SIP contact name is currently hardcoded to be a random token. It would be helpful if it was configurable, so that we can identify contacts in our telephony servers more easily. Asterisk for example shows this in the output of `pjsip show contacts` command. Our idea is to use username from our webapp + short random token (in case of multiple registrations).
There seems to be a comment indicating this is planned: https://github.com/onsip/SIP.js/blob/36a6bc2c6dbb22a54dc72cf3795c5efe28ca8c3c/src/api/user-agent.ts#L630
So this is kind of a reminder for version 16.

**Describe the solution you'd like**
A simple string property in UserAgentOptions.

**Describe alternatives you've considered**
Currently using something like this this before starting UA: `ua.contact.uri.user = sipContactName`. This is obviously terrible, and probably wouldn't even work if we finally started using typescript.

**Comment by john-e-riordan** [comment]
Yup.

And I concur that your alternative solution is the least worst alternative - before calling `ua.start()`, do something like `ua.contact.uri.user = sipContactName`.

While the piece of code you referenced can be addressed quickly by adding a configuration option (and perhaps we should do that now), it is the tip of a small iceberg which I’d like to attempt to address all at once after 0.16 goes out. It’s a particularly troublesome iceberg that I believe/hope we will have an opportunity rid ourselves of in a relatively painless way at that time. Barring objection, I’m fine with quickly adding an option, let’s say `contactUser`, knowing we are making said iceberg a wee bit bigger in the near term as it is a relatively tiny addition and clearly in the right direction. I’m also happy to punt it down the road a bit.

In short, with regard to the iceberg, the current “contact code” is currently both exposed to API consumer while simultaneously reaching all the way down into the bowels of the core message generation code in what is a kinda shocking top to bottom abstraction violation. Unfortunately this has resulted in years of existing library users deploying the same general approach you are currently taking to solve all sorts of contact related issues while simultaneously preventing some updates to the library even down at very bottom of the code stack without potentially breaking peoples applications in unexpected and difficult to fix ways. My thinking is once the library has been unburdened from the legacy API code (and the concern around breaking years of who know what contact hacking) we can update the core to “fix it right” while also providing everyone a route forward to the new API using a hopefully healthy abstraction.

What follows are some poorly organized and incomplete notes on the topic (for myself if nobody else) to help get the head around the extent of the issue…

The tip of the iceberg is the following code which is now lives in `user-agent.ts` which has been carried and ported forward basically intact from the original jssip fork…

```
/**
   * Initialize contact.
   */
  private initContact(): Contact {
    const contactName = createRandomToken(8); // FIXME: should be configurable
    const contactTransport =
      this.options.hackWssInTransport ? "wss" : "ws"; // FIXME: clearly broken for non ws transports
    const contact = {
      pubGruu: undefined,
      tempGruu: undefined,
      uri: new URI("sip", contactName, this.options.viaHost, undefined, { transport: contactTransport }),
      toString: (contactToStringOptions: { anonymous?: boolean, outbound?: boolean } = {}) => {
        const anonymous = contactToStringOptions.anonymous || false;
        const outbound = contactToStringOptions.outbound || false;
        let contactString: string = "<";
        if (anonymous) {
          contactString += this.contact.tempGruu || `sip:anonymous@anonymous.invalid;transport=${contactTransport}`;
        } else {
          contactString += this.contact.pubGruu || this.contact.uri;
        }
        if (outbound) {
          contactString += ";ob";
        }
        contactString += ">";
        return contactString;
      }
    };
    return contact;
  }
```
A stab getting some kind of isolation around the code continuing to infect the core was made here…

```
/**
 * Contact.
 * @remarks
 * https://tools.ietf.org/html/rfc3261#section-8.1.1.8
 * This is ported from UA.contact.
 * FIXME: TODO: This is not a great rep for Contact
 * and is used in a kinda hacky way herein.
 * @public
 */
export interface Contact {
  pubGruu: URI | undefined;
  tempGruu: URI | undefined;
  uri: URI;
  toString: (options?: { anonymous?: boolean, outbound?: boolean }) => string;
}
```
There are effected classes/interfaces at the api layer including…
```
Registerer
RegistererOptions
UserAgent 
UserAgentOptions
```
… where there are a bunch of configuration options that get tied in including…
```
RegisterOptions.extraContactHeaderParams
RegisterOptions.instanceId
RegisterOptions.regId
UserAgentOptions.hackIpInContact
UserAgentOptions.hackViaTcp
UserAgentOptions.hackWssInTransport
UserAgentOptions.viaHost
```
… and also down in the core all this get linked into classes/interfaces including…
```
UserAgentCore
UserAgentCoreConfiguration
OutgoingRequestMessage
OutgoingRequestMessageOptions
```
See the first line of `OutgoingRequestMessage.setViaHeader()` for fun example.

**Comment by inoric** [comment]
Yeah, took me quite a while to figure out where the contact name is set, so I did see some of the mess. Really sticks out compared to how clean the rest of the new API is.
At least most of that is now hidden behind the pretty API for most use cases.

If you think this is something people often need to do, maybe add some kind of "contactOptions" or something like that now to discourage this kind of workaround, to cause less breakage later when cleaning it up. If not, at least I'm using "the least worst alternative".

**Comment by egreenmachine** [comment]
Fixed in 9a9413702015437b1e0abe6226afc1c1b6bb69ea
